bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieNettyServer.java

```diff
@@ -112,19 +151,121 @@ boolean isRunning() {
     void suspendProcessing() {
         synchronized (suspensionLock) {
             suspended = true;
-            allChannels.setReadable(false).awaitUninterruptibly();
+            for (Channel channel : allChannels) {
+                channel.config().setAutoRead(false);
+            }
         }
     }
 
     @VisibleForTesting
     void resumeProcessing() {
         synchronized (suspensionLock) {
             suspended = false;
-            allChannels.setReadable(true).awaitUninterruptibly();
+            for (Channel channel : allChannels) {
+                channel.config().setAutoRead(true);
+            }
             suspensionLock.notifyAll();
         }
     }
 
+    private void listenOn(InetSocketAddress address, BookieSocketAddress bookieAddress) throws InterruptedException {
+        if (!conf.isDisableServerSocketBind()) {
```

Comment:
For reasons of code coupling, correctness, and code reuse I really would prefer to see us refactor the code.  I think all of this can be turned into something like.

```
private void listenOn(InetSocketAddress address, BookieSocketAddress bookieAddress) throws InterruptedException {
    if (eventLoopGroup != null) {
        listenOn(eventLoopGroup, address, bookieAddress);
    }
    if (jvmEventLoopGroup != null) {
        listenOn(jvmEventLoopGroup, address, bookieAddress);
    }
}

private void listenOn(EventLoopGroup elg, InetSocketAddress address, BookieSocketAddress bookieAddress) throws InterruptedException {
            ServerBootstrap bootstrap = new ServerBootstrap();
            bootstrap.childOption(ChannelOption.ALLOCATOR, new PooledByteBufAllocator(true));
            bootstrap.group(eventLoopGroup, eventLoopGroup);
            bootstrap.childOption(ChannelOption.TCP_NODELAY, conf.getServerTcpNoDelay());
            bootstrap.childOption(ChannelOption.SO_LINGER, conf.getServerSockLinger());
            bootstrap.childOption(ChannelOption.RCVBUF_ALLOCATOR,
                    new AdaptiveRecvByteBufAllocator(conf.getRecvByteBufAllocatorSizeMin(),
                            conf.getRecvByteBufAllocatorSizeInitial(), conf.getRecvByteBufAllocatorSizeMax()));

            if (elg instanceof DefaultEventLoopGroup) {
                bootstrap.channel(LocalServerChannel.class);
            } else if (elg instanceof EpollEventLoopGroup) {
                bootstrap.channel(EpollServerSocketChannel.class);
            } else if (elg instanceof NioEventLoopGroup) {
                jbootstrap.channel(NioServerSocketChannel.class);
            } else {
                throw new IllegalArgumentException("Don't know how to handle an EventLoopGroup of type " + elg.getClass());
            }

            bootstrap.childHandler(new BookieChannelInitializer());
            if (elg instanceof DefaultEventLoopGroup) {
                bootstrap.bind(bookieAddress.getLocalAddress()).sync();
                LocalBookiesRegistry.registerLocalBookieAddress(bookieAddress);
            } else {
                // Bind and start to accept incoming connections
                bootstrap.bind(address.getAddress(), address.getPort()).sync();
            }
}

...

class BookieChannelInitializer extends ChannelInitializer<Channel> {
        @Override
        public void initChannel(Channel ch) throws Exception {
            synchronized (suspensionLock) {
                while (suspended) {
                    suspensionLock.wait();
                }
            }
            BookieSideConnectionPeerContextHandler contextHandler = new BookieSideConnectionPeerContextHandler();
            ChannelPipeline pipeline = ch.pipeline();
            pipeline.addLast("lengthbaseddecoder",
                    new LengthFieldBasedFrameDecoder(maxMessageSize, 0, 4, 0, 4));
            pipeline.addLast("lengthprepender", new LengthFieldPrepender(4));

            pipeline.addLast("bookieProtoDecoder", requestDecoder);
            pipeline.addLast("bookieProtoEncoder", responseEncoder);
            pipeline.addLast("bookieAuthHandler",
                    new AuthHandler.ServerSideHandler(contextHandler.getConnectionPeer(), authProviderFactory));

            ChannelInboundHandler requestHandler = isRunning.get() ? new BookieRequestHandler(conf,
                    requestProcessor, allChannels) : new RejectRequestHandler();

            pipeline.addLast("bookieRequestHandler", requestHandler);
            pipeline.addLast("contextHandler", contextHandler);
        }
}

```